File: connectors/sdk/redact.py
"""Runner-side secret scrub (job spec §7, security review #1 F3 / D-66
point 2).

A connector exception can echo a resolved credential — a driver whose
error message includes the DSN, a traceback line carrying it. This module
scrubs exception messages and tracebacks by pattern *before* they are
placed into a ``JobError`` (runner.py), so no resolved secret travels the
`fail` wire call into `jobs.error` / `health_events.detail`. The core runs
the same pass again as defense in depth (core/src/redact.ts).

Pattern-based, not value-based: connection URIs with userinfo
(``postgres://user:pass@host/db``), libpq keyword secrets (``password=…``),
and bearer tokens. A match collapses to a fixed marker.
"""

import re
from typing import Any

REDACTION_MARKER = "[redacted:credential]"

_CREDENTIAL_URI = re.compile(r"\b[a-z][a-z0-9+.\-]*://[^\s:/@]+:[^\s/@]+@\S+", re.IGNORECASE)
_KEYWORD_SECRET = re.compile(
    r"\b(password|passwd|pwd|secret|token|api[_-]?key)\s*=\s*(\"[^\"]*\"|'[^']*'|\S+)",
    re.IGNORECASE,
)
_BEARER = re.compile(r"\bBearer\s+[A-Za-z0-9._\-]+", re.IGNORECASE)


def redact_text(value: str) -> str:
    """Redact credential-shaped substrings in one string (no-op when none)."""
    value = _CREDENTIAL_URI.sub(REDACTION_MARKER, value)
    value = _KEYWORD_SECRET.sub(lambda m: f"{m.group(1)}={REDACTION_MARKER}", value)
    value = _BEARER.sub(REDACTION_MARKER, value)
    return value


def redact_deep(value: Any) -> Any:
    """Recursively redact every string in a JSON-ish value; structure kept."""
    if isinstance(value, str):
        return redact_text(value)
    if isinstance(value, list):
        return [redact_deep(v) for v in value]
    if isinstance(value, tuple):
        return tuple(redact_deep(v) for v in value)
    if isinstance(value, dict):
        return {k: redact_deep(v) for k, v in value.items()}
    return value

File: core/src/redact.ts
/**
 * Defense-in-depth secret redaction over stored error detail (job spec
 * §7, security review #1 finding F3 / ruling D-66 point 2). The runner
 * scrubs its own exception messages and tracebacks before delivery
 * (connectors/sdk/redact.py); this is the core-side belt-and-suspenders
 * pass, applied to anything about to land in `jobs.error` or
 * `health_events.detail` — the two rows any bearer-token holder can read.
 *
 * It matches credential-shaped strings by pattern (the core never holds
 * resolved secret values — J-4 — so it cannot match by value): connection
 * URIs carrying userinfo (`postgres://user:pass@host/db`), libpq-style
 * keyword credentials (`password=…`), and bearer tokens. Matches collapse
 * to a fixed marker so the redaction is visible in triage, never the value.
 */

export const REDACTION_MARKER = "[redacted:credential]";

/** URI with embedded credentials: scheme://user:secret@host/… */
const CREDENTIAL_URI = /\b[a-z][a-z0-9+.\-]*:\/\/[^\s:/@]+:[^\s/@]+@\S+/gi;
/** libpq/connection-string keyword secrets: password=…, token=…, api_key=… */
const KEYWORD_SECRET =
  /\b(password|passwd|pwd|secret|token|api[_-]?key)\s*=\s*("[^"]*"|'[^']*'|\S+)/gi;
/** Bearer credentials that may ride in a copied header/detail. */
const BEARER = /\bBearer\s+[A-Za-z0-9._\-]+/gi;

/** Redact credential-shaped substrings in one string. No-op when none. */
export function redactString(value: string): string {
  return value
    .replace(CREDENTIAL_URI, REDACTION_MARKER)
    .replace(KEYWORD_SECRET, (_m, key: string) => `${key}=${REDACTION_MARKER}`)
    .replace(BEARER, REDACTION_MARKER);
}

/** Recursively redact every string in a JSON-ish value; structure preserved. */
export function redactDeep<T>(value: T): T {
  if (typeof value === "string") return redactString(value) as unknown as T;
  if (Array.isArray(value)) return value.map((v) => redactDeep(v)) as unknown as T;
  if (value && typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value)) out[k] = redactDeep(v);
    return out as T;
  }
  return value;
}

/** Redact an error envelope's message + detail before storage (§7). */
export function redactError<T extends { message?: string; detail?: unknown }>(error: T): T {
  return {
    ...error,
    ...(typeof error.message === "string" ? { message: redactString(error.message) } : {}),
    ...(error.detail !== undefined ? { detail: redactDeep(error.detail) } : {}),
  };
}
